src/modm/driver/inertial/qmc5883l.hh

```diff
@@ -0,0 +1,231 @@
+
+#include <modm/architecture/interface/i2c_device.hpp>
+#include <modm/architecture/interface/register.hpp>
+
+#include <modm/math/utils/endianness.hpp>
+
+template <typename I2cMaster>
+class Qmc5883l;
+
+struct Qmc5883lRegisters
+{
+protected:
+	/// @cond
+	/// The addresses of the Configuration and Data Registers
+	enum class
+	Register : uint8_t
+	{
+		DataX_Lsb = 0x00,
+		DataX_Msb = 0x01,
+		DataY_Lsb = 0x02,
+		DataY_Msb = 0x03,
+		DataZ_Lsb = 0x04,
+		DataZ_Msb = 0x05,
+    Status    = 0x06,
+    Tout_Lsb  = 0x07,
+    Tout_Msb  = 0x08,
+    Control1  = 0x09,
+    Control2  = 0x0a,
+    SetReset  = 0x0b,
+	};
+	/// @endcond
+public:
+  enum class
+  Status : uint8_t
+  {
+    DOR =  modm::Bit2,
+    OVL =  modm::Bit1,
+    DRDY = modm::Bit0,
+  };
+  MODM_FLAGS8(Status);
+
+public:
+  enum class
+  Control1 : uint8_t
+  {
+    OSR1 = modm::Bit7,
+    OSR0 = modm::Bit6,
+    OversampleRate_Mask = OSR1 | OSR0,
+
+    RNG1 = modm::Bit5,
+    RNG0 = modm::Bit4,
+    FullScale_Mask = RNG1 | RNG0,
+
+    ODR1 = modm::Bit3,
+    ODR0 = modm::Bit2,
+    OutputDataRate_Mask = ODR1 | ODR0,
+
+    MODE1 = modm::Bit1,
+    MODE0 = modm::Bit0,
+    Mode_Mask = MODE1 | MODE0,
+  };
+  MODM_FLAGS8(Control1);
+
+  enum class
+  Control2 : uint8_t
+  {
+    SOFT_RST = modm::Bit7,
+    ROL_PNT  = modm::Bit6,
+    INT_ENB  = modm::Bit0,
+  };
+  MODM_FLAGS8(Control2);
+
+public:
+  enum class
+  OversampleRate : uint8_t
+  {
+    _512 = 0,
+    _256 = int(Control1::OSR0),
+    _128 = int(Control1::OSR1),
+    _64  = int(Control1::OSR0) | int(Control1::OSR1),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, OversampleRate);
+
+  enum class
+  FullScale : uint8_t
+  {
+    _2G = 0,
+    _8G = int(Control1::RNG0),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, FullScale);
+
+  enum class
+  OutputDataRate : uint8_t
+  {
+    _10Hz  = 0,
+    _50Hz  = int(Control1::ODR0),
+    _100Hz = int(Control1::ODR1),
+    _200Hz = int(Control1::ODR0) | int(Control1::ODR1),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, OutputDataRate);
+
+  enum class
+  Mode : uint8_t
+  {
+    StandBy = 0,
+    Continious = int(Control1::MODE0),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, Mode);
+
+public:
+	struct modm_packed
+  Data
+	{
+		template <class I2cMaster>
+		friend class Qmc5883l;
+  protected:
+    uint8_t data[7];
+
+    template<Register R, uint8_t Offset = uint8_t(R)>
+    int16_t inline getWord() {
+      static_assert(Offset < sizeof data);
+      const auto value = reinterpret_cast<int16_t*>( data + Offset );
+      return modm::fromLittleEndian( *value );
+    }
+  };
+};
+
+
+template <class I2cMaster>
+class Qmc5883l : public Qmc5883lRegisters, public modm::I2cDevice<I2cMaster>
+{
+public:
+  // This flag is supposed to be set from within an ISR. By the time when the
+  // first call to readData() occurs in the program, the DRDY bit will have
+  // been already set, so need to initialize this flag to "true" to enable
+  // first data read.
+  volatile bool dataReady = true;
```

Comment:
The idea behind this flag was to provide an automatic storage which, one way or another, has to be allocated somewhere. Without it a typical use case would involve one variable declaration for the class instance and one variable declaration for the flag. So even if the "ready" statuc is polled, the mechanic stays the same - set `dataReady` to _true_, call `readBlocking`. If, with this information in mind, you still don't want to see it - I'll remove it. Please, let me know.

Reply:
I appreciate the idea, however, you're mixing drivers with scheduling without knowing exactly how the application schedules things. Maybe a delay is required because multiple readings need to be synchronized, maybe its easier to poll the pin input trigger flag, rather than to use an interrupt to prevent an interrupt storm. It's much easier to just delegate this to the application layer and let the driver be "dumb". In addition, `Blocking` has been used in modm to wrap `RF_CALL_BLOCKING`, rather than to model scheduler timings. So just implement the synchronization in the example with something like `if (flag.exchange(false)) { driver.read(); }`.